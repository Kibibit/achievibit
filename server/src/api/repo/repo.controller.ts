import {
  Controller,
  Logger,
  NotFoundException,
  Param,
  UseFilters
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { GetAll, GetOne } from '@kb-decorators';
import { KbValidationExceptionFilter } from '@kb-filters';
import { Repo } from '@kb-models';

import { RepoService } from './repo.service';

@Controller('api/repo')
@ApiTags('repo')
@UseFilters(new KbValidationExceptionFilter())
export class RepoController {
  private readonly logger = new Logger(RepoController.name);

  constructor(private readonly repoService: RepoService) {}

  @GetAll(Repo)
  async getAllRepos() {
    const repos = await this.repoService.findAllRepos();

    return repos;
  }

  @GetOne(Repo, ':name')
  async getRepo(@Param('name') name: string) {
    const repo = await this.repoService.findByName(name);

    if (!repo) {
      throw new NotFoundException(`Repo with name ${ name } not found`);
    }

    // will show secret fields as well!
    this.logger.log('Full Repo');
    // will log only public fields!
    this.logger.log(repo);
    // DANGER! WILL LOG EVERYTHING!
    // console.log(repo);

    // will only include exposed fields
    return repo;
  }
}
