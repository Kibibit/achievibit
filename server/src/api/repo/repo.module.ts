import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';

import { Repo } from '@kb-models';

import { RepoController } from './repo.controller';
import { RepoService } from './repo.service';

@Module({
  imports: [
    MongooseModule.forFeature([
      { name: Repo.modelName, schema: Repo.schema }
    ])
  ],
  providers: [RepoService],
  controllers: [RepoController],
  exports: [RepoService]
})
export class RepoModule {}
