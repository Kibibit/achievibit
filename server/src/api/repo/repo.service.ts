import { ReturnModelType } from '@typegoose/typegoose';

import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';

import { BaseService } from '@kb-abstracts';
import { Repo } from '@kb-models';

@Injectable()
export class RepoService extends BaseService<Repo> {
  constructor(
    @InjectModel(Repo.modelName)
    private readonly repoModel: ReturnModelType<typeof Repo>
  ) {
    super(repoModel, Repo);
  }

  async findAllRepos(): Promise<Repo[]> {
    const dbRepos = await this.findAll().exec();

    return dbRepos.map((repo) => new Repo(repo.toObject()));
  }

  async findByName(name: string): Promise<Repo> {
    const dbRepo = await this.findOne({ name }).exec();

    if (!dbRepo) {
      return;
    }

    return new Repo(dbRepo.toObject());
  }
}
