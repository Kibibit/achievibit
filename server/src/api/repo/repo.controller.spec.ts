import { noop } from 'lodash';

import { Test, TestingModule } from '@nestjs/testing';

import { RepoService } from '@kb-api';
import { DtoMockGenerator } from '@kb-dev-tools';

import { RepoController } from './repo.controller';

describe('RepoController', () => {
  let controller: RepoController;
  let repoService: RepoService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({ 
      controllers: [RepoController],
      providers: [{
        provide: RepoService,
        useValue: { findAllRepos: noop, findByName: noop }
      }]
    }).compile();

    controller = module.get<RepoController>(RepoController);
    repoService = module.get<RepoService>(RepoService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should call service on get all repos', async () => {
    const findAllResponse = DtoMockGenerator.repos();

    // console.log('mock result: ', findAllResponse);

    const spyFindAll = jest.spyOn(repoService, 'findAllRepos')
      .mockImplementation(() => Promise.resolve(findAllResponse));

    const result = await controller.getAllRepos();

    expect(result).toEqual(findAllResponse);
    expect(spyFindAll).toHaveBeenCalledTimes(1);
  });

  it('should get repo by name', async () => {
    const findByNameResponse = DtoMockGenerator.repo();

    // console.log('mock result: ', findAllResponse);
    const spyFindByUsername = jest.spyOn(repoService, 'findByName')
      .mockImplementation(() => Promise.resolve(findByNameResponse));

    const result = await controller.getRepo(findByNameResponse.name);

    expect(result).toEqual(findByNameResponse);
    expect(spyFindByUsername).toHaveBeenCalledTimes(1);
  });

  it('should throw error if repo not found', async () => {
    const spyFindByName = jest.spyOn(repoService, 'findByName')
      .mockImplementation(() => Promise.resolve(undefined));

    expect(controller.getRepo('mockname'))
      .rejects.toThrowErrorMatchingSnapshot();
    expect(spyFindByName).toHaveBeenCalledTimes(1);
  });
});
