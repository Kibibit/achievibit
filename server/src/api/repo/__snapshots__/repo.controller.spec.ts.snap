// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoController should throw error if repo not found 1`] = `"Repo with name mockname not found"`;
